### ISSUE TITLE
How to open application with parameters? 

### ISSUE THREAD

**Issue by rafaelfndev** [louaySleman/react-native-launcher-kit#7]
I need to open application with parameters, is possible? I looked in the documentation and there is a method to open applications, but I didn't find anything related to parameters:

`RNLauncherKitHelper.launchApplication('com.example.louay')`

I'm trying to open an emulator and specify the game it needs to open directly, how can I do this?

Example code:

```js
import { Button } from 'react-native';
import { RNLauncherKitHelper } from 'react-native-launcher-kit';

const App = () => {
  const packageName = 'com.explusalpha.NeoEmu';
  const gamePath = '/storage/emulated/0/MyFiles/Games/NeoGeo/kof98.zip';

  const openEmulatorWithGame = async () => {
    try {
      await RNLauncherKitHelper.launchApplication(packageName, {
        rom: gamePath, // Is possible something like that?
      });
    } catch (error) {
      console.error('Error:', error);
    }
  };

  return (
    <Button title="Open kof98 with NEO.emu" onPress={openEmulatorWithGame} />
  );
};

export default App;
```

**Comment by louaySleman** [comment]
Hello I'm so happy to tell you that starting from v2.0.0, you can pass parameters to the application you want to launch! Here's how you can do it:

```typescript
RNLauncherKitHelper.launchApplication('com.example.louay', {
 game: 'gameName',
 mode: 'gameMode',
 level: '5'
});
```
The second parameter accepts an object with any key-value pairs that the target application is configured to handle.

Let me know if you need any clarification on how to structure your parameters! 😊

@rafaelfndev

**Comment by rafaelfndev** [comment]
> Hello I'm so happy to tell you that starting from v2.0.0, you can pass parameters to the application you want to launch! Here's how you can do it:
> 
> ```ts
> RNLauncherKitHelper.launchApplication('com.example.louay', {
>  game: 'gameName',
>  mode: 'gameMode',
>  level: '5'
> });
> ```
> 
> The second parameter accepts an object with any key-value pairs that the target application is configured to handle.
> 
> Let me know if you need any clarification on how to structure your parameters! 😊
> 
> @rafaelfndev

Thank you very much for this update! 

Could you update your package on [NPM](https://www.npmjs.com/package/react-native-launcher-kit)?

Thanks =)

**Comment by louaySleman** [comment]
You're very welcome! 😊 The latest version is now live! 🚀 Check it out here: [react-native-launcher-kit v2.0.0](https://www.npmjs.com/package/react-native-launcher-kit/v/2.0.0) 🔗

@rafaelfndev

**Comment by rafaelfndev** [comment]
Hi, @louaySleman . 

I'm trying to open some applications by passing parameters, but none of them are working. Maybe I'm passing the data incorrectly. Could you check or provide some examples with real applications like Google Maps, Waze, or some Game Emulators?

For example, I'm trying to open **Google Maps** in a specific location:

```js
const openMap = async () => {
  const latitude = '-23.5692361';
  const longitude = '-46.6793614';

  const packageName = 'com.google.android.apps.maps';
  const geoUri = `geo:${latitude},${longitude}`;

  RNLauncherKitHelper.launchApplication(packageName, {
    data: geoUri,
  });
};
```

Or in this case where I'm trying to open a game with the **ePSXe emulator**, and even though the log shows that the file was found, it doesn't start the game directly:

```js
const openGame = async () => {
  const packageName = 'com.epsxe.ePSXe';

  // ROM file
  const gamePath = '/sdcard/MyFiles/ROMs/Play Station 1/Twisted Metal 4/Twisted Metal 4 (USA).cue';

  try {
    // Check if file exists
    const fileExists = await RNFS.exists(gamePath);

    if (fileExists) {
      console.log('File found:', gamePath);

      // Open ePSXe with ROM
      RNLauncherKitHelper.launchApplication(packageName, {
        data: `file://${gamePath}`,
      });
    } else {
      console.log('File not found:', gamePath);
    }
  } catch (error) {
    console.error('Error on open file:', error);
  }
};
```

Or trying to open a game in **MD.Emu**, passing some parameters, but it didn't work either:

```js
const openGame = async game => {
  const packageName = 'com.explusalpha.MdEmu';
  const gamePath = `file:///storage/emulated/0/MyFiles/ROMs/MegaDrive/${game}`;

  try {
    const fileExists = await RNFS.exists(gamePath);
    if (fileExists) {
      console.log('File found:', gamePath);
      RNLauncherKitHelper.launchApplication(packageName, {
        action: 'android.intent.action.VIEW',
        data: gamePath,
      });
    } else {
      console.log('File not found:', gamePath);
    }
  } catch (error) {
    console.error('Error on open file:', error);
  }
};
```

Or even to open a simple URL in **Google Chrome**:

```js
const openBrowser = () => {
  const packageName = 'com.android.chrome';
  const url = 'https://gmail.com/';
  RNLauncherKitHelper.launchApplication(packageName, {
    data: url,
  });
};
```

I looked for some ways to pass parameters to the intent, so I found this way with the "**extras**" key, but an error occurs when I try to pass the parameters (attached below code).

```js
const openGameAltMethod = async game => {
  const packageName = 'com.explusalpha.MdEmu';
  const gamePath = `file:///storage/emulated/0/MyFiles/ROMs/MegaDrive/${game}`;

  try {
    const fileExists = await RNFS.exists(gamePath);
    if (fileExists) {
      console.log('File found:', gamePath);
      RNLauncherKitHelper.launchApplication(packageName, {
        data: gamePath,
        extras: {
          "android.intent.extra.STREAM": gamePath,
        }
      });
    } else {
      console.log('File not found:', gamePath);
    }
  } catch (error) {
    console.error('Error on open file:', error);
  }
};
```

"Value for extras cannot be cast from ReadableNativeMap to String":

![Screenshot_1731453624](https://github.com/user-attachments/assets/a2a4c70f-4f84-49bf-b65e-5474946d4d48)

**Comment by louaySleman** [comment]
Hi @rafaelfndev ,
Thanks for reaching out! I'll take a look at your issue and provide examples or updates on how to pass parameters correctly to applications like Google Maps, Waze, or Game Emulators. I'll update you as soon as possible! 😊

**Comment by louaySleman** [comment]
Hi @rafaelfndev
Just released version 2.1.0 with enhanced launch parameters support and improved type safety. Would love to hear your feedback on the improvements! Feel free to reach out if you need any help with the implementation. 😊

**Comment by rafaelfndev** [comment]
Hi, @louaySleman . I tested the parameters you implemented and they worked very well in some applications, thank you very much for this implementation!

However, with the emulators (which is my main idea, to create a launcher for games), it is still not working. It opens when I do not pass any parameters, but if I try to pass any parameters, it simply does not open.

I recorded a video to show the behavior:
https://youtu.be/fd87UDuuMnQ

You can test by downloading the emulators from the developer's own page, they are free:
https://github.com/Rakashazi/emu-ex-plus-alpha

If you think this is not a problem with react-native-launcher-kit, I can try to talk to the emulator developer, but these same emulators work to open games directly when they are used with frontend managers, such as [Beacon](https://play.google.com/store/apps/details?id=com.radikal.gamelauncher), [Pegasus](https://pegasus-frontend.org/) or [Launchbox](https://www.launchbox-app.com/)

If you want to try, here os my code:

```js
import React from 'react';
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native';

import { IntentAction, RNLauncherKitHelper } from 'react-native-launcher-kit';

const App = () => {
  const item = ({ title, onPress }) => {
    return (
      <TouchableOpacity onPress={onPress}>
        <View style={[styles.box]}>
          <Text style={[styles.title]}>{title}</Text>
        </View>
      </TouchableOpacity>
    );
  };

  return (
    <View style={[styles.wrap]}>
      {item({
        title: 'Maps with geo',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.google.android.apps.maps', {
            action: IntentAction.VIEW,
            data: `geo:40.7580,-73.9855?q=-23.556418956888518, -46.68660045767081&z=16`,
          });
        },
      })}

      {item({
        title: 'Maps with navigation',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.google.android.apps.maps', {
            action: IntentAction.VIEW,
            data: `google.navigation:q=-23.556418956888518, -46.68660045767081&mode=driving`,
          });
        },
      })}

      {item({
        title: 'Chrome with Youtube',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.android.chrome', {
            action: IntentAction.VIEW,
            data: 'https://www.youtube.com',
          });
        },
      })}

      {item({
        title: 'MD.Emu (no game)',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.explusalpha.MdEmu');
        },
      })}

      {item({
        title: 'MD.Emu ACTIVITY',
        onPress: async () => {
          const result = await RNLauncherKitHelper.checkIfPackageInstalled('com.android.settings');
          console.log(result); // true or false
          RNLauncherKitHelper.launchApplication('com.explusalpha.MdEmu', {
            action: IntentAction.VIEW,
            data: 'file:///storage/emulated/0/md-game.bin',
            type: 'application/x-genesis-rom',
          });
        },
      })}

      {item({
        title: 'MD.Emu with game (type 1)',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.explusalpha.MdEmu', {
            action: IntentAction.VIEW,
            data: 'file:///storage/emulated/0/md-game.bin',
            type: 'application/x-genesis-rom',
          });
        },
      })}

      {item({
        title: 'MD.Emu with game (type 2)',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.explusalpha.MdEmu', {
            action: IntentAction.VIEW,
            data: 'file:///storage/emulated/0/md-game.bin',
            type: 'application/octet-stream',
          });
        },
      })}

      {item({
        title: 'NeoGeo KOF 97',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.expIusalpha.NeoEmu', {
            action: IntentAction.VIEW,
            data: 'file:///storage/emulated/0/kof97.zip',
          });
        },
      })}

      {item({
        title: 'NeoGeo (no ROM)',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.expIusalpha.NeoEmu');
        },
      })}

      {item({
        title: 'Messages',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('com.google.android.apps.messaging');
        },
      })}

      {item({
        title: 'AdAway',
        onPress: () => {
          RNLauncherKitHelper.launchApplication('org.adaway');
        },
      })}
    </View>
  );
};

const styles = StyleSheet.create({
  wrap: {
    gap: 10,
    padding: 20,
  },
  box: {
    flexDirection: 'column',
    borderWidth: 1,
    padding: 10,
    borderRadius: 5,
    backgroundColor: 'rgb(207, 233, 236)',
  },
  title: {
    fontWeight: 'bold',
    fontSize: 16,
  },
  description: {
    color: '#666',
    marginTop: 2,
  },
});

export { App };

```
